src/lib/storage/dictionary_column.hpp

```diff
@@ -0,0 +1,144 @@
+    _assign_attribute_vector(base_column->size());
+    _build_attribute_vector(base_column);
+  }
+
+  // return the value at a certain position. If you want to write efficient operators, back off!
+  const AllTypeVariant operator[](const size_t i) const override { return get(i); }
+
+  // return the value at a certain position.
+  const T get(const size_t i) const { return _dictionary->at(_attribute_vector->get(i)); }
+
+  // dictionary columns are immutable
+  void append(const AllTypeVariant&) override { throw std::logic_error("dictionary columns are immutable"); }
+
+  // returns an underlying dictionary
+  std::shared_ptr<const std::vector<T>> dictionary() const { return _dictionary; }
+
+  // returns an underlying data structure
+  std::shared_ptr<const BaseAttributeVector> attribute_vector() const { return _attribute_vector; }
+
+  // return the value represented by a given ValueID
+  const T& value_by_value_id(ValueID value_id) const { return _dictionary->at(value_id); }
+
+  // returns the first value ID that refers to a value >= the search value
+  // returns INVALID_VALUE_ID if all values are smaller than the search value
+  ValueID lower_bound(T value) const { return _to_value_id(value); }
+
+  // same as lower_bound(T), but accepts an AllTypeVariant
+  ValueID lower_bound(const AllTypeVariant& value) const { return _to_value_id(type_cast<T>(value)); }
+
+  // returns the first value ID that refers to a value > the search value
+  // returns INVALID_VALUE_ID if all values are smaller than or equal to the search value
+  ValueID upper_bound(T value) const {
+    const auto found = std::upper_bound(_dictionary->cbegin(), _dictionary->cend(), value);
+    if (found == _dictionary->cend()) {
+      return INVALID_VALUE_ID;
+    }
+    return ValueID(found - _dictionary->cbegin());
+  }
+
+  // same as upper_bound(T), but accepts an AllTypeVariant
+  ValueID upper_bound(const AllTypeVariant& value) const { return upper_bound(type_cast<T>(value)); }
+
+  // return the number of unique_values (dictionary entries)
+  size_t unique_values_count() const { return _dictionary->size(); }
+
+  // return the number of entries
+  size_t size() const override { return _attribute_vector->size(); }
+
+ protected:
+  std::shared_ptr<std::vector<T>> _dictionary;
+  std::shared_ptr<BaseAttributeVector> _attribute_vector;
+
+ protected:
+  void _build_dictionary(const std::shared_ptr<const BaseColumn>& base_column) {
+    _copy_values(base_column);
+    std::sort(_dictionary->begin(), _dictionary->end());
+    _dictionary->erase(std::unique(_dictionary->begin(), _dictionary->end()), _dictionary->end());
+    _dictionary->shrink_to_fit();
+  }
+
+  void _copy_values(const std::shared_ptr<const BaseColumn>& base_column) {
+    const auto value_column = std::dynamic_pointer_cast<const ValueColumn<T>>(base_column);
+    if (value_column) {
+      _copy_values_fast(value_column);
+    } else {
+      _copy_values_slow(base_column);
```

Comment:
Does it make sense to copy the values of a non-value column (e.g., compress a dictionary column)?

Reply:
Actually we do not know about which column types are going to be added. I did not want to neglect the fact that other column types may exist in the system. I think it is meaningful to provide a slow-but-working fallback implementation instead of restricting ourselves to VCs early on.